### PR TITLE
Refactor podman system connection

### PR DESCRIPTION
**Pull request by jwhonce**
* Add support to manage multiple connections
  * Add connection
  * Remove connection
  * Rename connection
  * Set connection as default

Signed-off-by: Jhon Honce <jhonce@redhat.com>

### CODE REVIEW

**Comment by baude** [comment]
@rhatdan wdyt?

**Comment by rhatdan** [comment]
I don't see the reason for rename, I would just tell them to edit the file.  Or Remove and re-add.

**Comment by jwhonce** [comment]
Blocked on https://github.com/containers/podman/pull/7000

**Comment by rhatdan** [comment]
#7000 merged.

**Comment by edsantiago** [comment]
Ouch - I see the reason for the test failure now. As it is now, this command breaks precedent because it is the only `podman something` subcommand that does not require an argument. All the others -- `image`, `container`, `pod`, `network`, whatever -- can never be invoked standalone, they all require a sub-subcommand. 

Super-unfortunately, this is nonintuitive because the `--help` message implies that it's optionsl:
```console
$ podman something --help
Usage:
  podman something [command]
```
One way around this is to special-case the `.bats` test so it recognizes and skips `podman system connection`.

Another way, which I lean toward, is to add `ls` so `podman system connection ls` will list connections, and with no args it will barf. This preserves consistency with every other podman subcommand; I also think it's a tad more intuitive. But it's your call.

**Comment by rhatdan** [comment]
I like the idea of ls

File: cmd/podman/system/connection/remove.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	"github.com/spf13/cobra"
+)
+
+var (
+	// Skip creating engines since this command will obtain connection information to said engines
+	rmCmd = &cobra.Command{
+		Use:                   "remove NAME",
+		Args:                  cobra.ExactArgs(1),
+		Aliases:               []string{"rm"},
+		Long:                  `Remove ssh destination information podman configuration`,
```
Comment (TomSweeneyRedHat) [inline cmd/podman/system/connection/remove.go:17]: should information be removed here?

Comment (TomSweeneyRedHat) [reply cmd/podman/system/connection/remove.go:17]: or maybe
```suggestion
		Long:                  `Remove ssh destination information for the podman configuration`,
```

File: cmd/podman/system/connection/rename.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+)
+
+var (
+	// Skip creating engines since this command will obtain connection information to said engines
+	renameCmd = &cobra.Command{
+		Use:                   "rename OLD NEW",
+		Aliases:               []string{"mv"},
+		Args:                  cobra.ExactArgs(2),
+		Short:                 "Rename remote ssh destination",
+		Long:                  `Rename ssh destination information name "old" to "new"`,
```
Comment (TomSweeneyRedHat) [inline cmd/podman/system/connection/rename.go:20]:
```suggestion
		Long:                  `Rename the ssh destination information name from "old" to "new"`,
```

File: docs/source/markdown/podman-system-connection-add.1.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
+% podman-system-connection-add(1)
+
+## NAME
+podman\-system\-connection\-add - Record ssh destination for podman service
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection-add.1.md:4]:
```suggestion
podman\-system\-connection\-add - Record ssh destination for the Podman service
```

File: docs/source/markdown/podman-system-connection-add.1.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+## OPTIONS
+
+**-d**, **--default**=*false*
+
+Make the new destination the default for this user.
+
+**--identity**=*path*
+
+Path to ssh identity file. If the identity file has been encrypted, podman prompts the user for the passphrase.
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection-add.1.md:24]:
```suggestion
Path to ssh identity file. If the identity file has been encrypted, Podman prompts the user for the passphrase.
```

File: docs/source/markdown/podman-system-connection-add.1.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+If no identity file is provided and no user is given, podman defaults to the user running the podman command.
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection-add.1.md:25]:
```suggestion
If no identity file is provided and no user is given, Podman defaults to the user running the podman command.
```

File: docs/source/markdown/podman-system-connection-add.1.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+Podman prompts for the login password on the remote server.
+
+**-p**, **--port**=*port*
+
+Port for ssh destination. The default value is `22`.
+
+**--socket-path**=*path*
+
+Path to podman service unix domain socket on the ssh destination host
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection-add.1.md:34]:
```suggestion
Path to the Podman service unix domain socket on the ssh destination host
```

File: docs/source/markdown/podman-system-connection-default.1.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+% podman-system-connection-default(1)
+
+## NAME
+podman\-system\-connection\-default - Set named ssh destination as default for podman service(s)
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection-default.1.md:4]:
```suggestion
podman\-system\-connection\-default - Set named ssh destination as default for Podman service(s)
```

File: docs/source/markdown/podman-system-connection-default.1.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+## SYNOPSIS
+**podman system connection default** *name*
+
+## DESCRIPTION
+Set named ssh destination as default destination for podman service(s).
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection-default.1.md:10]:
```suggestion
Set named ssh destination as default destination for Podman service(s).
```

File: docs/source/markdown/podman-system-connection-list.1.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
+% podman-system-connection-list(1)
+
+## NAME
+podman\-system\-connection\-list - List ssh destination(s) for podman service
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection-list.1.md:4]:
```suggestion
podman\-system\-connection\-list - List ssh destination(s) for the Podman service
```

File: docs/source/markdown/podman-system-connection-remove.1.md [hunk 10]
```diff
@@ -0,0 +1,20 @@
+% podman-system-connection-remove(1)
+
+## NAME
+podman\-system\-connection\-remove - Delete named ssh destination
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection-remove.1.md:4]: Delete or Remove?

Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection-remove.1.md:4]:
```suggestion
podman\-system\-connection\-remove - Delete the named ssh destination for the Podman service
```

File: docs/source/markdown/podman-system-connection-rename.1.md [hunk 11]
```diff
@@ -0,0 +1,20 @@
+% podman-system-connection-rename(1)
+
+## NAME
+podman\-system\-connection\-rename - Rename ssh destination for podman service
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection-rename.1.md:4]:
```suggestion
podman\-system\-connection\-rename - Rename ssh destination for the Podman service
```

File: docs/source/markdown/podman-system-connection-rename.1.md [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+## SYNOPSIS
+**podman system connection rename** *old* *new*
+
+## DESCRIPTION
+Rename ssh destination from *old* to *new*.
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection-rename.1.md:10]:
```suggestion
Rename the ssh destination from *old* to *new*.
```

File: docs/source/markdown/podman-system-connection.1.md [hunk 13]
```diff
@@ -1,43 +1,34 @@
 % podman-system-connection(1)
 
 ## NAME
-podman\-system\-connection - Record ssh destination for remote podman service
+podman\-system\-connection - Manage ssh destination for podman service(s)
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection.1.md:4]:
```suggestion
podman\-system\-connection - Manage ssh destination for the Podman service(s)
```

File: docs/source/markdown/podman-system-connection.1.md [hunk 14]
```diff
@@ -1,43 +1,34 @@
...
 
 ## SYNOPSIS
-**podman system connection** [*options*] [*ssh destination*]
+**podman system connection** *subcommand*
 
 ## DESCRIPTION
-Record ssh destination for remote podman service(s). The ssh destination is given as one of:
- - [user@]hostname[:port]
- - ssh://[user@]hostname[:port]
+Manage ssh destination(s) for podman service(s).
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection.1.md:10]:
```suggestion
Manage ssh destination(s) for the Podman service(s).
```

File: docs/source/markdown/podman-system-connection.1.md [hunk 15]
```diff
@@ -1,43 +1,34 @@
...
 
-The user will be prompted for the remote ssh login password or key file pass phrase as required. `ssh-agent` is supported if it is running.
+The user will be prompted for the ssh login password or key file pass phrase as required. `ssh-agent` is supported if it is running.
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection.1.md:12]:
```suggestion
The user will be prompted for the ssh login password or key file pass phrase as required. The `ssh-agent` is supported if it is running.
```

File: docs/source/markdown/podman-system-connection.1.md [hunk 16]
```diff
@@ -1,43 +1,34 @@
...
-**--socket-path**=*path*
-
-Path to podman service unix domain socket on the ssh destination host
+| Command  | Man Page                                                                     | Description                                                |
+| -------  | ---------------------------------------------------------------------------- | ---------------------------------------------------------- |
+| add      | [podman-system-connection-add(1)](podman-system-connection-add.1.md)         | Record ssh destination for podman service |
+| default  | [podman-system-connection-default(1)](podman-system-connection-default.1.md) | Set named ssh destination as default for podman service(s) |
+| list     | [podman-system-connection-list(1)](podman-system-connection-list.1.md)       | List ssh destination(s) for podman service |
+| remove   | [podman-system-connection-remove(1)](podman-system-connection-remove.1.md)   | Delete named ssh destination |
+| rename   | [podman-system-connection-rename(1)](podman-system-connection-rename.1.md)   | Rename ssh destination for podman service |
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection.1.md:22]: if you take changes from above, reflect them here too

File: docs/source/markdown/podman-system.1.md [hunk 17]
```diff
@@ -11,17 +11,16 @@ The system command allows you to manage the podman systems
...
+| connection | [podman-system-connection(1)](podman-system-connection.1.md) | Manage ssh destination for podman service(s)                         |
+| df         | [podman-system-df(1)](podman-system-df.1.md)                 | Show podman disk usage.                                              |
+| info       | [podman-system-info(1)](podman-info.1.md)                    | Displays Podman related system information.                          |
+| migrate    | [podman-system-migrate(1)](podman-system-migrate.1.md)       | Migrate existing containers to a new podman version.                 |
+| prune      | [podman-system-prune(1)](podman-system-prune.1.md)           | Remove all unused container, image and volume data.                  |
+| renumber   | [podman-system-renumber(1)](podman-system-renumber.1.md)     | Migrate lock numbers to handle a change in maximum number of locks.  |
+| reset      | [podman-system-reset(1)](podman-system-reset.1.md)           | Reset storage back to initial state.                                 |
+| service    | [podman-system-service(1)](podman-system-service.1.md)       | Run an API service                                                   |
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system.1.md:23]: ditto if you take changes from above, here too.

File: docs/source/markdown/podman-system-connection-add.1.md [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+
+## SYNOPSIS
+**podman system connection add** [*options*] *name* *destination*
+
+## DESCRIPTION
+Record ssh destination for remote podman service(s). The ssh destination is given as one of:
+ - [user@]hostname[:port]
+ - ssh://[user@]hostname[:port]
+
+The user will be prompted for the remote ssh login password or key file pass phrase as required. `ssh-agent` is supported if it is running.
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-system-connection-add.1.md:14]:
```suggestion
The user will be prompted for the remote ssh login password or key file pass phrase as required. The `ssh-agent` is supported if it is running.
```

File: cmd/podman/root.go [hunk 19]
```diff
@@ -236,16 +236,12 @@ func loggingHook() {
...
 	lFlags := cmd.Flags()
-	custom, _ := config.ReadCustomConfig()
-	defaultURI := custom.Engine.RemoteURI
-	if defaultURI == "" {
-		defaultURI = registry.DefaultAPIAddress()
-	}
 	lFlags.BoolVarP(&opts.Remote, "remote", "r", false, "Access remote Podman service (default false)")
-	lFlags.StringVar(&opts.URI, "url", defaultURI, "URL to access Podman service (CONTAINER_HOST)")
-	lFlags.StringVar(&opts.Identity, "identity", custom.Engine.RemoteIdentity, "path to SSH identity file, (CONTAINER_SSHKEY)")
+	lFlags.StringVar(&opts.URI, "url", uri, "URL to access Podman service (CONTAINER_HOST)")
```
Comment (edsantiago) [inline cmd/podman/root.go:243]: I'm feeling pretty squicked out by the interchange of URI and URL. I think that will lead to maintenance headaches. From my (limited) understanding, I believe URL (l as in Lima) is the correct term. Would you consider replacing the code instances of `uri` (i as India) to `url` for consistency?

File: docs/source/markdown/podman-system-connection.1.md (hunk 15 above)
Comment (edsantiago) [reply docs/source/markdown/podman-system-connection.1.md]: Day late dollar short, I know (you've already made the change) - but I've never heard anyone use an article to refer to ssh-agent. It looks and sounds super weird to read 'the ssh-agent`. But, maybe that's just me.

File: cmd/podman/root.go (hunk 19 above)
Comment (rhatdan) [reply cmd/podman/root.go:243]: Yes let's stick with UR**L**

Comment (jwhonce) [reply cmd/podman/root.go:243]: uri was chosen for a reason. url is a golang package name. There is a lot of code that uses the url package to process the "uri" variable.  We would be forced to alias the package name or make up longer names which I deemed to be a large maintenance headache.  I attempt to use uri for internals and URL for user facing so everyone is happy.

As a url is a uri this naming simplifies the  coding and is correct naming.

**Comment by rhatdan** [comment]
LGTM

File: cmd/podman/root.go (hunk 19 above)
Comment (mheon) [reply cmd/podman/root.go:243]: Would calling it `path` be a good compromise? It's less specific as to what exact type of path it is, and only one character longer.

Comment (jwhonce) [reply cmd/podman/root.go:243]: Path is an element of the url/uri, so that would be confusing to see path.Path.

Comment (rhatdan) [reply cmd/podman/root.go:243]: How about we call everything URI and drop URL, if we want to still support the --url for backwards compatibility just make it an alias.

**Pull request merged**